src/libstore/local-store.cc

```diff
@@ -997,18 +999,16 @@ void LocalStore::addToStore(const ValidPathInfo & info, Source & source,
 
             deletePath(realPath);
 
-            if (info.ca != "" &&
-                !((hasPrefix(info.ca, "text:") && !info.references.count(info.path))
-                    || info.references.empty()))
+            if (info.ca && !info.references.empty() && !std::holds_alternative<TextHash>(*info.ca))
```

Comment:
I think this is missing the `&& !info.references.count(info.path)` from the original condition when it's a text hash. That is, text hashes are allow but only when it has no self references.

Reply:
Good catch
```suggestion
            bool refersToSelf = info.references.count(info.path) > 0;
            if (info.ca && !info.references.empty() && !(std::holds_alternative<TextHash>(*info.ca) && !refersToSelf))
```